wordpress-mobile/WordPress-Android
# [Platform] Generate PlayStore Screenshots on CI

Fixes main part of #12601 

⚠️  This PR must NOT be merged as is, as it contains patches that I had to apply in order to make it work on my fork.
⚠️  ~This PR is built on top of #12831 and should only be merged after that one.~ [EDIT: #12831 is now merged]

## Scope

This PR implements the main part of #12601, i.e. making the WPAndroid raw and promo screenshots run on CI and Firebase Test Lab.

### Content of this PR

* Update the UI Tests to use FTL's ScreenShotter library in addition of still supporting `screengrab`.
* Add a `generate_screenshots` parameter to the `.circleci/config.yml` to trigger the `Screenshots` workflow on demand
* This workflow then starts a `raw-screenshots` job on a Linux executor (needed to use `gcloud`, `gsutil` and FTL)
   * This job builds the project then push it to Firebase Test Lab to run the UI Tests on a matrix of devices and locales.
   * Those UI Tests generate the raw screenshots and store them in a Firebase Cloud Storage bucket
   * The job then downloads the generated raw screenshots from FCS, then persists the bare minimum (`Gemfile`, `Gemfile.lock` and `fastlane/` folder) into the CI workspace.
* The workflow then subsequently runs a `promo-screenshots` job on a Mac executor (macOS needed to use `rmagick` and `drawText`)
   * This job restores the workspace saved by `raw-screenshots`. This avoids doing a full git clone of the source code while all we need is `bundler`, `fastlane`, and the raw screenshots in `fastlane/screenshots/raw`, not the source code
   * Then it installs ImageMagick and rmagick
   * Finally it updates the promo strings from GlotPress, Create the Promo Screenshots from the raw screenshots and `playstoreres` resources, and ZIP the result

The PR also adds multiple `fastlane` lanes:

* One to trigger the pipeline on CI from the CLI, intended to be the one you call during release cut
* One to download the raw screenshots from FCS and the `gs://` bucket, and reorganise them in a decent folder hierarchy. This one is mainly used by CI itself, but could also be triggered manually for debugging
* Changes in the `create_promo_screenshots` to convert the raw screenshots from JPG (format used by FTL) to PNG (format expected by fastlane) and rename them appropriately. (That lane still continues to work with screengrab-generated raw screenshots too)

## ⚠️ PR Dependency

~This PR is build on top of #12831 (fix for the crash in `ar` locale and stubs). It should only be merged once #12831 is merged.~

~To see only the partial diff (i.e. what the diff will look like after #12831 has been merged / what the diff should have looked like if I had been able to make this PR target the branch in my fork), [click here](https://github.com/AliSoftware/WordPress-Android/compare/feature/screenshots-phase1..AliSoftware:fix/stubs-ar-locale)~

[EDIT: Dependant PR #12831 has now been merged.]

## ⚠️ Draft State

Before merging, this PR needs to be changed in the following ways by a maintainer after importing the branch into the base repo:

### Revert patches used for fork / free FTL instance

Commits with the 🚩 emoji must be reverted. Those commits are only present in the current branch in order to make the build pass and work on my fork. Those include:

 - 🚩 Update of the Firebase Secrets JSON : Since I'm using my free-plan Firebase Instance, had to update the secrets JSON file and the project ID. This change should NOT be included in the main repo.
 - 🚩 Reducing the number of locales from 21 to 2, because my free plan on Firebase don't allow more than a 2x2 matrix at once. We likely want to restore running on all locales once integrating this PR in the base repo.
 - 🚩 Pointing to the dev release of the android Orb. Once the Orb is released via https://github.com/wordpress-mobile/circleci-orbs/pull/61 we should point to `@1.0` again

### Checking URLs and Tokens

* We need to ensure the `SLACK_WEBHOOK` env var we have in the main repo's CircleCI instance points to the appropriate channel where we want to post the "WPAndroid Screenshots are ready" message. (If not, we might need to amend the Slack step accordingly to post into the right channel)
* We also need to ensure the URLs used in the Slack messages are correct. Especially, those point to URLs that use some unique IDs in their host name, and those unique IDs typically depend on the repo CircleCI was building for


## To test:

### Test that screengrab still works

* Run `bundle exec fastlane  screenshots` to generate raw screenshots **locally** using _screengrab_
* Run `bundle exec fastlane create_promo_screenshots` to generate the promo screenshots from the screengrab-generated raw screenshots, and ensure they still work and look ok

### Test that the new FTL remote screenshots works

* Ensure we have a `SLACK_WEBHOOK` env var in the CI which points to the right channel
* Run `be fastlane trigger_screenshots token:<citoken> fork:AliSoftware branch:feature/screenshots-phase1`
   * If you already have a `CIRCLECI_TOKEN` env var defined locally, no need to pass the `token:<citoken>` parameter
   * If you want to test on the main repo instead of the fork, you can omit the `fork:` parameter
* Ensure the pipeline has started on the CircleCI instance (the one associated with the `fork:` you used)
* Once the CI pipeline has finished, check that:
   * The success message was received in the appropriate Slack channel
   * The link to download the ZIP file from the artefacts is valid
   * The content of the ZIP contains the expected screenshots with the appropriate, up-to-date and translated promo strings for each locale

## Next Steps

Once this PR gets merged and the `feature/run-screenshots-on-ci` main branch gets merged into `master`, we need to [update the Field Guide page](https://fieldguide.automattic.com/updating-screenshots-for-wpandroid/) accordingly.

---

## PR submission checklist:
